modules/core/shared/src/main/scala/io/circe/JsonObject.scala

```diff
@@ -275,17 +275,21 @@ object JsonObject {
    * Note that the order of the fields is arbitrary.
    */
   final def fromMap(map: Map[String, Json]): JsonObject =
-    if (map.isEmpty) {
-      empty
-    } else {
-      fromMapAndVector(map, map.keys.toVector)
-    }
+    fromMapAndVector(map, map.keys.toVector)
 
   private[circe] final def fromMapAndVector(map: Map[String, Json], keys: Vector[String]): JsonObject =
-    new MapAndVectorJsonObject(map, keys)
+    (map.size: @switch) match {
+      case 0 => empty
+      case 1 => new SingletonJsonObject(map.keys.iterator.next(), map.values.iterator.next())
+      case _ => new MapAndVectorJsonObject(map, keys)
+    }
 
   private[circe] final def fromLinkedHashMap(map: LinkedHashMap[String, Json]): JsonObject =
-    new LinkedHashMapJsonObject(map)
+    (map.size: @switch) match {
```

Comment:
Also this.

Reply:
For a Java Map, I don't think I can do better ? `size` is cached, and `keySet`/`values` are lazy